### ISSUE TITLE
Support cross-tenant authentication for Shared Image Gallery

### ISSUE THREAD

**Issue by haokanga** [terraform-providers/terraform-provider-azurerm#3548]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Carnegie Mellon University needs to be able to create VMs from images in the Azure Shared Image Gallery across tenants, as in this scenario: <https://docs.microsoft.com/en-us/azure/virtual-machines/linux/share-images-across-tenants>
And we would like to be able to do this via Terraform.

Teaching assistants/staff at CMU need to be able to publish an image to SIG (in the CMU domain) and let students (**in other domains**) create a VM from this image using Terraform.

### New or Affected Resource(s)

To do this, terraform will need to support `x-ms-authorization-auxiliary` tokens in the header - <https://docs.microsoft.com/en-us/azure/azure-resource-manager/authenticate-multi-tenant>

### References

<https://docs.microsoft.com/en-us/azure/virtual-machines/linux/share-images-across-tenants>
<https://docs.microsoft.com/en-us/azure/azure-resource-manager/authenticate-multi-tenant>
Here's an example of creating a VM from a SIG resource in ARM (not cross-tenant): https://github.com/Azure/azure-quickstart-templates/blob/master/101-vm-from-sig/azuredeploy.json

**Comment by nicethomaslearngit** [comment]
Hi All

I have created shared image gallery in tenant1 and I am able to create VMs using this image across all subscriptions in this tenant. Used CLI, powershell and Terraform and it worked fine

When I try to create a VM in Tenant-2 with Shared Image Galley available in Tenant-1, I am facing some issues. I am able to create the VM in tenant by using CLI and power shell, but when using Terraform it is failing with below error.

Error: compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=403 -- Original Error: Code="LinkedAuthorizationFailed" Message="The client has permission to perform action 'Microsoft.Compute/galleries/images/versions/read' on scope '/subscriptions/*aaaaaaaaaaaaaaaa/resourceGroups/mygroup/providers/Microsoft.Compute/virtualMachines/sigvm-01', however the current tenant 'bbbbbbbbbbbbbbb' is not authorized to access linked subscription 'ccccccccccccccccccccc'."

aaaaaaaaaaaaaa -> is the subscription ID in Tenant-2 where I am trying to create VM

bbbbbbbbbbbbbb -> is the tenant ID of tenant 2

cccccccccccccccc  ->  is the subscription ID in tenant1

ANY help is much appreciated.

**Comment by tombuildsstuff** [comment]
@nicethomaslearngit this functionality isn't yet supported (this issue's tracking support for it) - but it's being worked on in #4290

**Comment by nicethomaslearngit** [comment]
@tombuildsstuff   Thanks for your prompt response.. So you saying currently only CLI and powershell is the option and terraform is not supported yet?

**Comment by tombuildsstuff** [comment]
Fixed via #4290

**Comment by tombuildsstuff** [comment]
@nicethomaslearngit correct - it's been merged and will be available in v1.34.0 of the Azure Provider once it's released

**Comment by nicethomaslearngit** [comment]
Hi @tombuildsstuff ,

I see the version 1.34.0 is released, still I am getting the error I pasted initially.   My Azure Shared Image galleries are in tenant-1 and I am trying to deploy a VM in tenant-2 using this SIG in tenant-1.

```terraform
####  connecting to tenant -2 ########
provider "azurerm" {
  subscription_id = "${var.subscription_id}"
  client_id       = "${var.client_id}"
  client_secret   = "${var.client_secret}"
  tenant_id       = "${var.tenant_id}"
}

#####  connecting to tenant -1 ######## Provider for fetching the SIG resource
provider "azurerm" {
  alias           = "sig-resources-id"
  subscription_id = "${var.sig_subscription_id}"
  client_id       = "${var.sig_client_id}"
  client_secret   = "${var.sig_client_secret}"
  tenant_id       = "${var.sig_tenant_id}"
}
```

using alias I fetches the image ID and use this ID to provision in tenant-2 and is not working. Any suggestions here?
